### PR TITLE
Add Monad Transformers Post

### PR DESCRIPTION
**Pull request by lmreis89**

### CODE REVIEW

File: _posts/2016-04-11-monad-transformers.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    (for {
+      company <- OptionT.fromOption(getCompany(companyName))
+      if company.employees map(_.id) contains employeeId
+      details <- OptionT.liftF(getDetails(employeeId))
+    } yield details.age).value
+  }
+```
+
+The only changes here when compared to the previous example is that we're now using the `OptionT.fromOption` function in the first case, and the `OptionT.liftF` function in the second one. The `fromOption` function creates an `OptionT` from an `Option` monad. It is internally wrapping the return of the `getCompany` function in a `Future.successful` call. The `liftF` function lifts any monad `F` into an `OptionT`. Internally, it is calling the `map` function from the `Future` monad and wrapping the returned `EmployeeWithDetails` in a `Some`. It is important to note that this is just an example and there are more monad transformers like `EitherT`, `ListT`, etc, in both cats and scalaz.
+
```
Comment (ragb) [inline _posts/2016-04-11-monad-transformers.md:126]: StateT also, which is mentioned many times and is very useful.

File: _posts/2016-04-11-monad-transformers.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+case class Company(companyName: String, employees: List[EmployeeWithoutDetails])
+
+trait SyncDBOps {
+  protected def getDetails(employeeId: String): Option[EmployeeWithDetails]
+  protected def getCompany(companyName: String): Option[Company]
+}
+```
+
+From this simple design, the following conclusions are derived:
```
Comment (jtjeferreira) [inline _posts/2016-04-11-monad-transformers.md:29]: ordered lists aren't being generated. add a newline here

File: _posts/2016-04-11-monad-transformers.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ 1. An `Employee` is either an `EmployeeWithDetails` or an `EmployeeWithoutDetails`
+ 2. A `Company` has a name and a list of employees without their details
+ 3. A `SyncDBOps` can fetch a `Company` by its name and an `EmployeeWithDetails` by its id.
+ 
+Suppose we want to create a new software layer on top of this API and expose a single function with the following specification:
```
Comment (jtjeferreira) [inline _posts/2016-04-11-monad-transformers.md:35]: and another newline here

File: _posts/2016-04-11-monad-transformers.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+With `OptionT`, if the `Future` returns `Failure` or the `Option` returns `None`, the function will immediatelly return with that value. Other than the `OptionT` apply function, only the `.value` function is called. This function transforms the `OptionT` monad back into a `Future[Option[Company]]`. By using `OptionT` we have removed our biggest issue in the previous example where we were changing the semantics of the application when composing functions that used two monads. You can get more details about the implementation of `OptionT` and other monad transformers by going to the cats documentation provided at the end of this post.
+
+## More use cases for monad transformation
+Monad transformers aren't only used when composing functions that return two monads in the form M[F[T]] where M and F are distinct monads like in our previous example of `Future[Option[Company]]`. Suppose the API was changed into the following two functions:
```
Comment (jtjeferreira) [inline _posts/2016-04-11-monad-transformers.md]: add ` to M[F[T]] and M and F

File: _posts/2016-04-11-monad-transformers.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      if company.employees map(_.id) contains employeeId
+      details <- OptionT.liftF(getDetails(employeeId))
+    } yield details.age).value
+  }
+```
+
+The only changes here when compared to the previous example is that we're now using the `OptionT.fromOption` function in the first case, and the `OptionT.liftF` function in the second one. The `fromOption` function creates an `OptionT` from an `Option` monad. It is internally wrapping the return of the `getCompany` function in a `Future.successful` call. The `liftF` function lifts any monad `F` into an `OptionT`. Internally, it is calling the `map` function from the `Future` monad and wrapping the returned `EmployeeWithDetails` in a `Some`. It is important to note that this is just an example and there are more monad transformers like `EitherT`, `ListT`, etc, in both cats and scalaz.
+
+## Conclusion
+I hope you enjoyed this post and feel like getting started with function composition using multiple monads. The code used for this post is available at [E.Near's Monad Transformers](TODO:PUT LINK).
```
Comment (jtjeferreira) [inline _posts/2016-04-11-monad-transformers.md]: https://github.com/enear/Monad-Transformers-Tutorial

**Pull request merged**